File: app/js/lib/algorithms/genetic_algorithm.js
const GeneticAlgorithm = {
  genetic_algorithm(arr, populationSize = 50, generations = 100, mutationRate = 0.1) {
    const targetSum = arr.reduce((a, b) => a + b, 0);
    let population = Array.from({ length: populationSize }, () => [...arr].sort(() => Math.random() - 0.5));

    for (let g = 0; g < generations; g++) {
      const fitness = population.map(ind => Math.abs(ind.reduce((a, b) => a + b, 0) - targetSum));
      const maxFit = Math.max(...fitness);
      const matingPool = population.map((ind, i) => ({ ind, fit: maxFit - fitness[i] }))
        .sort((a, b) => b.fit - a.fit).slice(0, populationSize / 2).map(p => p.ind);

      const nextGeneration = [...matingPool];
      while (nextGeneration.length < populationSize) {
        const parent1 = matingPool[Math.floor(Math.random() * matingPool.length)];
        const parent2 = matingPool[Math.floor(Math.random() * matingPool.length)];
        const crossoverPoint = Math.floor(Math.random() * parent1.length);
        const child = [...parent1.slice(0, crossoverPoint), ...parent2.slice(crossoverPoint)];
        const mutated = child.map(x => Math.random() < mutationRate ? arr[Math.floor(Math.random() * arr.length)] : x);
        nextGeneration.push(mutated);
      }
      population = nextGeneration;
    }
    return population.reduce((best, ind) => {
      const indScore = Math.abs(ind.reduce((a, b) => a + b, 0) - targetSum);
      const bestScore = Math.abs(best.reduce((a, b) => a + b, 0) - targetSum);
      return indScore < bestScore ? ind : best;
    }, population[0]);
  }
};

module.exports = GeneticAlgorithm;

File: app/lib/algorithms/genetic_algorithm.rb
# frozen_string_literal: true

module GeneticAlgorithm
  module_function

  def genetic_algorithm(arr, population_size = 50, generations = 100, mutation_rate = 0.1)
    target_sum = arr.sum
    population = Array.new(population_size) { arr.shuffle }

    generations.times do
      fitness = population.map { |ind| (ind.sum - target_sum).abs }
      max_fit = fitness.max
      mating_pool = population.zip(fitness).sort_by { |_, f| max_fit - f }.map(&:first).take(population_size / 2)

      next_generation = mating_pool.dup
      while next_generation.length < population_size
        parent1, parent2 = mating_pool.sample(2)
        crossover_point = rand(parent1.length)
        child = parent1[0...crossover_point] + parent2[crossover_point..]
        child = child.map { |x| rand < mutation_rate ? arr.sample : x }
        next_generation << child
      end
      population = next_generation
    end

    population.min_by { |ind| (ind.sum - target_sum).abs }
  end
end
